_posts/2019-07-30-microfrontends-holidaycheck-part2.md

```diff
@@ -0,0 +1,168 @@
+
+* **Javascript**: max. 32 kB minified (~10kb to 15kb gzipped) 
+* **CSS**: max. 15 kB minified
+* **HTML**: max 10 kB
+* **Server Side Rendering**: P95 must be below 10ms
+
+These budgets are worth nothing if we wouldn't measure them and alert ourselves when we exceed them.
+To measure the size of the JavaScript bundles we build, we use the [webpack max asset size feature](https://webpack.js.org/configuration/performance/#performancemaxassetsize).
+You can add this to your webpack build with just a few lines of configuration:
+
+````js
+{
+  performance: {
+    hints: 'error', // Fail the build when asset size is exceeded
+    maxEntrypointSize: 32768,
+    assetFilter: function(assetFilename) {
+      return assetFilename.endsWith('.js');
+    },
+  }
+}
+````
+
+Additionally we measure the size of the JS, CSS and HTML with [Calibre](https://calibreapp.com). Calibre
+is a service which lets you define a schedule for Google Lighthouse and displays the results over time.
+Calibre also has performance budgets built in, so you can define the limits and get alerted when a budget is exceeded.
+We have setup Calibre to alert into a Slack channel. The team behind Calibre is super responsive and motivated when
+we need help, report bugs or suggest features, I strongly recommend everyone to try their service.
+
+<figure>
+    <figcaption>Overview of a test run with Calibre</figcaption>
+    <img src="/img/posts/2019-07-30-microfrontends-holidaycheck-part2/calibre.png" alt="calibre-overview" />
+</figure>
+
+## Analysing
+For deeper analysis we use Google Lighthouse together with the various capabilities of the Chrome Dev Tools.
+Explaining the Chrome Dev Tools would exceed the scope of the article. I recommend everyone to get familiar
+with the Dev Tools. When it comes to speed especially the Network, Performance and Audits tabs are relevant.
+
+## Bundle Size
+In the today's JS ecosystem it is quite a challenge to build an App with a modern component based approach
+within 32 kB. Considering React + React-Dom have already 110 kB minified.
+
+That's why we decided to use Preact + Unistore, which together are roughly 10 kB minified. Preact has the same
+API as React with a similar feature set (besides synthetic events), but is *a lot* smaller. Unistore is the
+complementary tiny state management solution.
+
+It still happened regularly that we hit the 32kB for our main bundle and our CI pipeline gently reminded
+us with a red build. To debug what takes up the budget the Webpack Bundle Analyzer plugin is really great.
+It generates an interactive tree map which let's you easily find the big chunks.
+
+<figure>
+    <figcaption>Tree Map generated by Webpack Bundle Analyzer</figcaption>
+    <img src="/img/posts/2019-07-30-microfrontends-holidaycheck-part2/analyze-bundle.png" alt="bundle-analyzer-tree-map" />
+</figure>
+
+### Bundle Splitting
+Once we found the problematic parts we needed to find a solution. The biggest effect had **bundle splitting**.
```

Comment:
Something like "**Bundle Splitting** was the one thing with the highest impact" could sound better here.